internal/overlord/logstate/syslog.go

```diff
@@ -0,0 +1,202 @@
+}
+
+// Syslog Priority values - see RFC 5424 6.2.1
+const (
+	// Facility values
+	FacilityUserLevelMessage = 1
+
+	// Severity values
+	SeverityInformational = 6
+)
+
+// priorityVal calculates the syslog Priority value (PRIVAL) from the given
+// Facility and Severity values. See RFC 5424, sec 6.2.1 for details.
+func priorityVal(facility, severity int) int {
+	return facility*8 + severity
+}
+
+// syslogStructuredData formats the given labels into a structured data section
+// for a syslog message, according to RFC5424 section 6.
+func syslogStructuredData(name string, enterpriseNum int, labels map[string]string) string {
+	if len(labels) == 0 {
+		return "-"
+	}
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "[%s@%d", name, enterpriseNum)
+	for key, value := range labels {
+		fmt.Fprintf(&buf, " %s=\"", key)
+		// escape the value according to RFC5424 6.3.3
+		for i := 0; i < len(value); i++ {
+			// don't use "for _, c := range value" as we don't want runes
+			c := value[i]
+			if c == '"' || c == '\\' || c == ']' {
+				buf.WriteByte('\\')
+			}
+			buf.WriteByte(c)
+		}
+		buf.WriteByte('"')
+	}
+	buf.WriteByte(']')
+	return buf.String()
+}
+
+func (b *syslogClient) encodeEntry(w io.Writer, entry servicelog.Entry) {
+	// format defined by RFC 5424
+	timestamp := entry.Time.Format(time.RFC3339)
+	msg := fmt.Sprintf("<%d>%d %s %s %s %s %s %s %s",
+		b.priority, b.version, timestamp, b.host, entry.Service,
+		b.pid, b.msgid, b.structuredData, entry.Message)
+
+	// Octet framing as per RFC 5425.
+	framed := fmt.Sprintf("%d %s", len(msg), msg)
+	w.Write([]byte(framed))
+}
+
+// Send messages to remote syslog server.
+func (s *syslogClient) Send(entries []servicelog.Entry) error {
+	err := s.ensureConnected()
+	if err != nil {
+		return err
+	}
+
+	data := &bytes.Buffer{}
+	for _, entry := range entries {
+		s.encodeEntry(data, entry)
+	}
+
+	_, err = io.Copy(s.conn, data)
+	if err != nil {
+		// The connection might be bad. Close and reset it for later reconnection attempt(s).
+		s.conn.Close()
+	}
+	return err
+}
+
+func (s *syslogClient) ensureConnected() error {
+	if s.conn != nil {
+		return nil
+	} else if s.closed {
+		return fmt.Errorf("write to closed SyslogBackend")
+	}
+
+	if s.waitReconnect > 0 {
+		time.Sleep(s.waitReconnect)
+	}
+
+	conn, err := net.Dial(s.address.Scheme, s.address.Host)
+	if err != nil {
+		// start an exponential backoff for reconnection attempts
+		if s.waitReconnect == 0 {
+			s.waitReconnect = 100 * time.Millisecond
+		}
+		newWait := 2 * s.waitReconnect
+		if newWait > 10*time.Second {
+			newWait = 10 * time.Second
+		}
+		s.waitReconnect = newWait
+		return err
+	}
+
+	if s.conn != nil {
+		err = s.conn.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	s.waitReconnect = 0 // reset backoff
+	s.conn = conn
+	return nil
+}
+
+func (s *syslogClient) Close() error {
+	s.closed = true
+	func() {
+		if s.conn != nil {
+			s.conn.SetDeadline(time.Now().Add(-1 * time.Second))
+		}
+	}()
+
+	if s.conn == nil {
```

Comment:
Might be simpler to put this early-return at the top, and then only do the deadline/close stuff following that.